connectors/sink-google-sheets/internal/sink.go

```diff
@@ -0,0 +1,144 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+	"strconv"
+	ce "github.com/cloudevents/sdk-go/v2"
+	cdkgo "github.com/linkall-labs/cdk-go"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+	"google.golang.org/api/sheets/v4"
+)
+
+var _ cdkgo.Sink = &GoogleSheetSink{}
+
+func NewGoogleSheetSink() cdkgo.Sink {
+	return &GoogleSheetSink{}
+}
+
+type GoogleSheetSink struct {
+	config *GoogleSheetConfig
+}
+
+func (s *GoogleSheetSink) Initialize(ctx context.Context, cfg cdkgo.ConfigAccessor) error {
+	// TODO
+	s.config = cfg.(*GoogleSheetConfig)
+	
+	return nil
+}
+
+func (s *GoogleSheetSink) Name() string {
+	// TODO
+	return "GoogleSheetSink"
+}
+
+func (s *GoogleSheetSink) Destroy() error {
+	// TODO
+	return nil
+}
+
+func (s *GoogleSheetSink) Arrived(ctx context.Context, events ...*ce.Event) cdkgo.Result {
+	// TODO
+	for _, event := range events {
+
+		s.saveDataToSpreadsheet(event)
+	}
+	return cdkgo.SuccessResult
+}
+
+
+
+func (s *GoogleSheetSink) saveDataToSpreadsheet(event *ce.Event) {
+
+		//Create API Context
+	ctx := context.Background()
+
+	// authenticate and get configuration
+	config, err := google.JWTConfigFromJSON([]byte(s.config.Credentials), "https://www.googleapis.com/auth/spreadsheets")
+		if err != nil {
+			log.Fatalf("Failed to authenticate google service accounts key %v", err)
+			return
+		}
+
+	//Create Client
+	client := config.Client(ctx)
+
+	//Create Service using Client
+	srv, err := sheets.NewService(ctx, option.WithHTTPClient(client))
+	if err != nil {
+        log.Fatalf("Failed to Create Service Account %v",err)
+        return
+	}
+
+	//Initialize Sheet ID & Spreadsheet ID
+	
+	spreadSheetUrl := s.config.Sheet_url
+	
+	sheetId, err := strconv.Atoi(spreadSheetUrl[93:94])
+	if err != nil {
+        log.Fatalf("Failed to Convert String %v",err)
+        return
+	}
+
+	spreadSheetId := spreadSheetUrl[39:83]
+
+	//Get SheetName from SpreadSheetID
+	res1, err := srv.Spreadsheets.Get(spreadSheetId).Fields("sheets(properties(sheetId,title))").Do()
+	if err != nil {
+        log.Fatalf("Failed to get SheetName %v",err)
+        return
+	}
+
+	sheetName := ""
+	for _, v := range res1.Sheets {
+		prop := v.Properties
+		if prop.SheetId == int64(sheetId) {
+			sheetName = prop.Title
+			break
+		}
+	}
+
+
+	sheetRow := make(map[string]string)
+	sheetRow["id"] = "ID"
+	sheetRow["name"] = "Name"
+	sheetRow["email"] = "Email"
+	sheetRow["description"] = "Description"
+	sheetRow["date"] = "Date"
+	
+	extractData := json.Unmarshal(event.Data(), &sheetRow)
+	if extractData != nil {
+		log.Fatalf("Unable to Unmarshal %v",extractData)
+	}
+	
+
+	//Insert Row Value
+	row := &sheets.ValueRange{
+		Values: [][] interface{}{{sheetRow["id"], sheetRow["name"], sheetRow["email"], sheetRow["description"], sheetRow["date"] }},
```

Comment:
you can't hard code use id, name, email..., maybe the event data doesn't have these params, such as my event like 
```json
{
  "specversion": "1.0",
  "id": "53d1c340-551a-11ed-96c7-8b504d95037c",
  "source": "quickstart",
  "type": "quickstart",
  "datacontenttype": "application/json",
  "time": "2022-10-26T10:38:29.345Z",
  "data": {
    "myData": "simulation event data"
  }
}
```
1. you need to get the sheet header which is the sheet's first row
2. then you will have an array header which may be like header = []string{"id", "name",...} 

when an event arrived
1. event.Data to a map
```go
var sheetRow map[stirng]interface{}
json.Unmarshal(event.Data(), &sheetRow)
```
2. make a sheet row value
```go
var values []interface
for _,key:= range header{
 v,ok:= sheertRow[key]
 if !ok{
   //todo event data doesn't has the key
 }
  values = append(values, v)
}
```
3. make a sheet ValueRange
```go
row := &sheets.ValueRange{
		Values: [][] interface{}{values}
}
```

Reply:
2. Array header is created first by user, not with code